File: limablu_mobile/lib/data/repository.dart
// lib/data/repository.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

import 'api_client.dart';
import 'models.dart';

class Repo {
  final ApiClient api;
  Repo(this.api);

  // Providers (static defaults for now; match your color map)
  List<ProviderInfo> defaultProviders() => [
        ProviderInfo(provider: "SolarOne", pricePerKwh: 0.12, capacityKwh: 500),
        ProviderInfo(provider: "GreenLight", pricePerKwh: 0.10, capacityKwh: 400),
        ProviderInfo(provider: "SunStream", pricePerKwh: 0.14, capacityKwh: 300),
        ProviderInfo(provider: "RayPower", pricePerKwh: 0.16, capacityKwh: 250),
        ProviderInfo(provider: "BrightSolar", pricePerKwh: 0.11, capacityKwh: 350),
        ProviderInfo(provider: "NovaEnergy", pricePerKwh: 0.13, capacityKwh: 280),
      ];

  // Users (backend if available, else fallback single user)
  Future<List<UserConfig>> fetchUsers() async {
    try {
      final list = await api.listUsers(); // GET /users -> List<Map>
      final users = list.map<UserConfig>((m) {
        final name = (m["name"] ?? "") as String;
        final profile = (m["profile"] ?? "basic_lighting") as String;
        final alphaNum = (m["alpha"] ?? m["price_priority"] ?? 0.5) as num;
        return UserConfig(name: name, profile: profile, alpha: alphaNum.toDouble());
      }).toList();
      return users;
    } catch (_) {
      return [UserConfig(name: "Amina", profile: "basic_lighting", alpha: 0.5)];
    }
  }

  // Optimize (delegate to ApiClient)
  Future<Map<String, dynamic>> optimize({
    required List<ProviderInfo> providers,
    required List<UserConfig> users,
    required int days,
  }) {
    return api.optimize(providers: providers, users: users, days: days);
  }

  // Bill (delegate to ApiClient)
  Future<Map<String, dynamic>> bill({
    required List<Map<String, dynamic>> alloc,
    required List<ProviderInfo> providers,
    required String model, // "postpaid"|"token"|"subscription"
    double? tokenKwh,
    double? allowanceKwh,
    bool rollover = true,
  }) {
    return api.bill(
      alloc: alloc,
      providers: providers,
      model: model,
      tokenKwh: tokenKwh,
      allowanceKwh: allowanceKwh,
      rollover: rollover,
    );
  }

  // Save user preference (price vs capacity)
  Future<Map<String, dynamic>> patchUserAlpha(String name, double alpha) async {
    try {
      return await api.patchUserAlpha(name, alpha);
    } catch (_) {
      // offline echo so UI updates
      return {"name": name, "price_priority": alpha, "offline": true};
    }
  }

  // Buy Power (used by BuyPowerScreen)
  Future<Map<String, dynamic>> buyPower(double amountKsh) {
    return api.buyPower(amountKsh: amountKsh);
  }
}

File: emappapp/emappa-mobile/src/api/repository.ts
import { apiClient } from './client';
import {
  ProviderInfo,
  UserConfig,
  OptimizeResponse,
  BillResponse,
  BuyPowerResponse,
} from '../types';

class Repository {
  // Default providers (match Flutter app)
  defaultProviders(): ProviderInfo[] {
    return [
      { provider: 'SolarOne', price_per_kwh: 0.12, capacity_kwh: 500 },
      { provider: 'GreenLight', price_per_kwh: 0.10, capacity_kwh: 400 },
      { provider: 'SunStream', price_per_kwh: 0.14, capacity_kwh: 300 },
      { provider: 'RayPower', price_per_kwh: 0.16, capacity_kwh: 250 },
      { provider: 'BrightSolar', price_per_kwh: 0.11, capacity_kwh: 350 },
      { provider: 'NovaEnergy', price_per_kwh: 0.13, capacity_kwh: 280 },
    ];
  }

  // Fetch users from backend, fallback to default
  async fetchUsers(): Promise<UserConfig[]> {
    try {
      const list = await apiClient.listUsers();
      return list.map((m) => ({
        name: m.name || '',
        profile: m.profile || 'basic_lighting',
        alpha: m.alpha ?? 0.5,
      }));
    } catch (e) {
      // Fallback single user
      return [{ name: 'Amina', profile: 'basic_lighting', alpha: 0.5 }];
    }
  }

  // Optimize allocation
  async optimize(
    providers: ProviderInfo[],
    users: UserConfig[],
    days: number
  ): Promise<OptimizeResponse> {
    return apiClient.optimize(providers, users, days);
  }

  // Get billing
  async bill(
    alloc: Array<Record<string, unknown>>,
    providers: ProviderInfo[],
    model: 'postpaid' | 'token' | 'subscription',
    tokenKwh?: number,
    allowanceKwh?: number,
    rollover: boolean = true
  ): Promise<BillResponse> {
    return apiClient.bill(alloc, providers, model, tokenKwh, allowanceKwh, rollover);
  }

  // Update user preference
  async patchUserAlpha(name: string, alpha: number): Promise<UserConfig> {
    try {
      const result = await apiClient.patchUserAlpha(name, alpha);
      return {
        name: result.name,
        profile: result.profile || 'basic_lighting',
        alpha: result.alpha ?? alpha,
      };
    } catch (e) {
      // Offline echo so UI updates
      return { name, profile: 'basic_lighting', alpha };
    }
  }

  // Buy power tokens
  async buyPower(amountKsh: number): Promise<BuyPowerResponse> {
    return apiClient.buyPower(amountKsh);
  }
}

// Singleton instance
export const repository = new Repository();
